File: src/main/java/pw/vodes/rimuru/rss/Feed.java
package pw.vodes.rimuru.rss;

import java.io.IOException;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import org.javacord.api.entity.channel.ServerTextChannel;
import org.javacord.api.entity.message.embed.EmbedBuilder;

import com.apptasticsoftware.rssreader.RssReader;
import com.fasterxml.jackson.annotation.JsonIgnore;

import pw.vodes.rimuru.Main;
import pw.vodes.rimuru.Util;

public class Feed {
	
	public String name, url, serverID, channelID;
	public String regex;
	public List<FeedItem> items;
	public boolean firstCheck = true;
	
	@JsonIgnore
	private ServerTextChannel channel;
	
	@JsonIgnore
	private Pattern pattern;
	
	
	public Feed() {}

	public Feed(String name, String url, String regex, String serverID, String channelID) {
		this.name = name;
		this.url = url;
		this.regex = regex;
		this.serverID = serverID;
		this.channelID = channelID;
		this.items = new ArrayList<>();
	}
	
	@JsonIgnore
	public ServerTextChannel getChannel() {
		if(channel == null) {
			channel = Main.api.getServerById(serverID).get().getTextChannelById(channelID).get();
		}
		return channel;
	}
	
	@JsonIgnore
	public Pattern getRegexPattern() {
		if(pattern == null && regex != null && !regex.isBlank())
			pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
		return pattern;
	}
	
	public void check() {
		var reader = new RssReader();
		var feedURL = url;
		if(feedURL.toLowerCase().contains("u2.dmhy.org")) {
			if(Main.getConfig().getU2PassKey().isBlank())
				return;
			var matcher = Util.u2PasskeyURLPattern.matcher(feedURL);
			if(matcher.find()) {
				var toReplace = matcher.group(1);
				feedURL = feedURL.replace(toReplace, "passkey=" + Main.getConfig().getU2PassKey());
			}
		}
		try {
			var read = reader.read(feedURL).collect(Collectors.toList());
			
			for(var i : read) {
				var feedItem = FeedItem.ofItem(i);
				if(!items.contains(feedItem)) {
					var pat = getRegexPattern();
					if(pat != null) {
						if(!feedItem.getTitle().isBlank()) {
							var matcher = pat.matcher(feedItem.getTitle());
							if(matcher.find()) 
								items.add(feedItem);
						}
					} else 
						items.add(feedItem);
				}
			}
			
			items.sort((o1, o2) -> Long.compare(o2.publicationUnixTime(), o1.publicationUnixTime()));
			if(items.size() > 50) {
				items = items.subList(0, 49);
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
		
		postUnposted();
		if(firstCheck) {
			for(var item : items) {
				item.setWasPosted(true);
			}
			firstCheck = false;
			FeedManager.save();
		}
	}
	
	private void postUnposted() {
		var isU2 = url.toLowerCase().contains("u2.dmhy.org");
		var isNyaa = url.toLowerCase().contains("nyaa.si");
		for(var item : items) {
			if(item.wasPosted())
				continue;
			var embed = new EmbedBuilder();
			embed.setTitle(item.getTitle());
			
			if(isU2 || isNyaa) {
				embed.setAuthor(isU2 ? "U2" : "Nyaa", isU2 ? "https://u2.dmhy.org" : "https://nyaa.si", isU2 ? "https://i.imgur.com/lNorPYS.png" : "https://nyaa.si/static/favicon.png");
				if(isNyaa && !item.getDescription().isBlank()) {
					var matcher = Util.nyaaAutismPattern.matcher(item.getDescription());
					if(matcher.find()) {
						embed.setDescription(String.format("%s | %s", matcher.group(4), matcher.group(5)));
					}
				}
			}
			
			var image = item.getImage();
			if(image != null) {
				embed.setImage(image);
			}
			
			var url = item.getURLToPost();
			if(!url.isBlank())
				embed.setUrl(url);
			
			embed.setTimestamp(Instant.ofEpochSecond(item.publicationUnixTime()));
			getChannel().sendMessage(embed).thenRun(() -> item.setWasPosted(true));
			if(firstCheck) {
				if(Instant.ofEpochSecond(item.publicationUnixTime()).isBefore(Instant.now().minus(24, ChronoUnit.HOURS))) {
					break;
				}
			}
		}
	}
}

File: src/main/kotlin/pw/vodes/rimurukt/services/rss/Feed.kt
package pw.vodes.rimurukt.services.rss

import com.apptasticsoftware.rssreader.RssReader
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import org.javacord.api.entity.channel.ServerTextChannel
import org.javacord.api.entity.message.embed.EmbedBuilder
import org.unbescape.html.HtmlEscape
import pw.vodes.rimurukt.Main
import pw.vodes.rimurukt.ctI
import pw.vodes.rimurukt.reportException
import java.time.Instant
import java.time.temporal.ChronoUnit

val BAD_NAMING_TAGS = listOf("smol", "koala", "legion")

@Serializable
data class Feed(val name: String, val url: String, val regex: String, val serverID: String, val channelID: String) {
    var items = mutableListOf<FeedItem>()
    var firstCheck = true

    @Transient
    private var _channel: ServerTextChannel? = null

    fun channel(): ServerTextChannel {
        if (_channel == null)
            _channel = Main.api.getServerById(serverID).get().getTextChannelById(channelID).get()
        return _channel as ServerTextChannel
    }

    private fun getURL(): String {
        var feedURL = url
        if (url.contains("u2.dmhy.org", true)) {
            if (Main.config.u2Passkey.isBlank())
                return feedURL
            val match = u2PasskeyURLPattern.find(url)
            if (match != null) {
                feedURL = feedURL.replace(match.groups[1]!!.value, "passkey=${Main.config.u2Passkey}")
            }
        }
        return feedURL
    }

    fun check() {
        val reader = RssReader()
        val feedURL = getURL()
        try {
            reader.read(feedURL).forEach { item ->
                try {
                    val feedItem = FeedItem.ofItem(item)
                    val existing = items.find { it == feedItem }
                    if (existing != null)
                        return@forEach
                    if (regex.isBlank()) {
                        items.add(feedItem)
                        return@forEach
                    }
                    val matches = regex.toRegex(RegexOption.IGNORE_CASE).matches(feedItem.title)
                    if (matches)
                        items.add(feedItem)
                } catch (ex: Exception) {
                    reportException(ex, "Parsing posts for $name")
                }
            }
            var sorted = items.sortedByDescending { it.getUnixPubTime() }
            if (sorted.size > 50)
                sorted = sorted.subList(0, 49)
            items = sorted.toMutableList()
        } catch (ex: Exception) {
            reportException(ex, "Retrieving posts for $name")
        }

        postNew()
        if (firstCheck) {
            items.forEach { it.wasPosted = true }
            firstCheck = false
        }
    }

    private fun postNew() {
        val isU2 = url ctI "u2.dmhy.org"
        val isNyaa = url ctI "nyaa.si"

        for (item in items) {
            if (item.wasPosted)
                continue
            if (firstCheck && Instant.ofEpochSecond(item.getUnixPubTime()).isBefore(Instant.now().minus(24, ChronoUnit.HOURS)))
                continue

            var title = HtmlEscape.unescapeHtml(item.title)
            var badNaming = false
            for (tag in BAD_NAMING_TAGS) {
                if (title.contains("[$tag]", true) || title.contains("-$tag", true))
                    badNaming = true
            }
            if (badNaming && isNyaa) {
                title = if (title.length > 256) title.substring(0, 239) else title
                title += " (Bad Naming)"
            } else
                title = if (title.length > 256) title.substring(0, 250) else title
            val embed = EmbedBuilder().setTitle(title)

            if (isU2 || isNyaa) {
                embed.setAuthor(
                    if (isU2) "U2" else "Nyaa",
                    if (isU2) "https://u2.dmhy.org" else "https://nyaa.si",
                    if (isU2) "https://i.imgur.com/lNorPYS.png" else "https://nyaa.si/static/favicon.png"
                )
                if (isNyaa) {
                    val match = nyaaParsePattern.find(item.description)
                    if (match != null)
                        embed.setDescription("${match.groups[4]!!.value} | ${match.groups[5]!!.value}")
                }
            }

            val image = item.getImage()
            if (!image.isNullOrBlank())
                embed.setImage(image)

            embed.setUrl(item.getPostURL())
            embed.setTimestamp(Instant.ofEpochSecond(item.getUnixPubTime()))
            val future = channel().sendMessage(embed).thenAccept {
                item.wasPosted = true
                it.crossPost()
            }.exceptionally {
                reportException(it, "Could not post FeedItem with URL: ${item.getPostURL()}")
                null
            }
            // I want this to be blocking so stuff gets saved properly. Not really sure of a better way to do it.
            try {
                future.join()
            } catch (_: Exception) {
            }
        }
    }
}